.build/run-tests-enhanced.sh

```diff
@@ -0,0 +1,454 @@
+  local -r test_name_regexp="${test_name_regexp:-}"
+
+  # check split_chunk is compatible with target (if not a regexp). TODO support repeats for these
+  if [[ "${_split_chunk}" =~ ^\d+/\d+$ ]] && [[ "1/1" != "${split_chunk}" ]] ; then
+    case ${target} in
+      "stress-test" | "fqltool-test" | "microbench" | "cqlsh-test")
+          echo "Target ${target} does not suport splits."
+          die
+          ;;
+        *)
+          ;;
+    esac
+  fi
+
+  pushd ${CASSANDRA_DIR}/ >/dev/null
+
+  # jdk check
+  local -r java_version=$(java -version 2>&1 | awk -F '"' '/version/ {print $2}' | awk -F. '{print $1}')
+  local -r version=$(grep 'property\s*name=\"base.version\"' build.xml |sed -ne 's/.*value=\"\([^"]*\)\".*/\1/p')
+  local -r java_version_default=`grep 'property\s*name="java.default"' build.xml |sed -ne 's/.*value="\([^"]*\)".*/\1/p'`
+
+  if [ "${java_version}" -eq 17 ] && [[ "${target}" == "jvm-dtest-upgrade" ]] ; then
+    echo "Invalid JDK${java_version}. Only overlapping supported JDKs can be used when upgrading, as the same jdk must be used over the upgrade path."
+    die
+  fi
+
+  # check project is already built. no cleaning is done, so jenkins unstash works, beware.
+  [[ -f "${DIST_DIR}/apache-cassandra-${version}.jar" ]] || [[ -f "${DIST_DIR}/apache-cassandra-${version}-SNAPSHOT.jar" ]] || { echo "Project must be built first. Use \`ant jar\`. Build directory is ${DIST_DIR} with: $(ls ${DIST_DIR})"; die; }
+
+  # check if dist artifacts exist, this breaks the dtests
+  [[ -d "${DIST_DIR}/dist" ]] && { echo "tests don't work when build/dist ("${DIST_DIR}/dist") exists (from \`ant artifacts\`)"; die; }
+
+  # ant test setup
+  export TMP_DIR="${DIST_DIR}/tmp"
+  [ -d ${TMP_DIR} ] || mkdir -p "${TMP_DIR}"
+  export ANT_TEST_OPTS="-Dno-build-test=true -Dtmp.dir=${TMP_DIR}"
+
+  # fresh virtualenv and test logs results everytime
+  [[ "/" == "${DIST_DIR}" ]] || rm -rf "${DIST_DIR}/test/{html,output,logs}"
+
+  # cheap trick to ensure dependency libraries are in place. allows us to stash only project specific build artifacts.
+  ant -quiet -silent resolver-dist-lib
+
+  case ${target} in
+    "stress-test")
+      # hard fail on test compilation, but dont fail the test run as unstable test reports are processed
+      ant stress-build-test ${ANT_TEST_OPTS}
+      ant $target ${ANT_TEST_OPTS} || echo "failed ${target} ${split_chunk}"
+      ;;
+    "fqltool-test")
+      # hard fail on test compilation, but dont fail the test run so unstable test reports are processed
+      ant fqltool-build-test ${ANT_TEST_OPTS}
+      ant $target ${ANT_TEST_OPTS} || echo "failed ${target} ${split_chunk}"
+      ;;
+    "microbench")
+      ant $target ${ANT_TEST_OPTS} -Dmaven.test.failure.ignore=true
+      ;;
+    "test")
+      _run_testlist "unit" "testclasslist" "${test_name_regexp}" "${split_chunk}" "$(_timeout_for 'test.timeout')"
+      ;;
+    "test-repeat")
```

Comment:
having to duplicate every test type with a `-repeat` suffix is a bit clumsy.
is there a cleaner way we can do this ? as it's only flipping out test_name_regexp and adding test count.